spec-0007/index.md

```diff
@@ -0,0 +1,152 @@
+
+### Concepts
+
+- `BitGenerator`: Generates a stream of pseudo-random bits. The default generator in NumPy (`numpy.random.default_rng`) uses PCG64.
+- `Generator`: Derives pseudo-random numbers from the bits produced by a `BitGenerator`.
+- `RandomState`: a [legacy object in NumPy](https://numpy.org/doc/stable/reference/random/index.html), similar to `Generator`, that produces random numbers based on the Mersenne Twister.
+
+### Constraints
+
+NumPy, SciPy, scikit-learn, scikit-image, and NetworkX all implement pseudo-random seeding in slightly different ways.
+Common keyword arguments include `random_state` and `seed`.
+In practice, the seed is also often controllable using `numpy.random.seed`.
+
+## Implementation
+
+Legacy behavior in packages such as scikit-learn (`sklearn.utils.check_random_state`) typically handle `None` (use the global seed state), an int (convert to `RandomState`), or `RandomState` object.
+
+Our recommendation here is a deprecation strategy which does not in _all_ cases adhere to the Hinsen principle[^hinsen],
+although it could very nearly do so by enforcing the use of `rng` as a keyword argument.
+
+[^hinsen]: The Hinsen principle states, loosely, that code should, whether executed now or in the future, return the same result, or raise an error.
+
+The [deprecation strategy](https://github.com/scientific-python/specs/pull/180#issuecomment-1515248009) is as follows.
+
+**Initially**, accept both `rng` and the existing `random_state`/`seed`/`...` keyword arguments.
+
+- If both are specified by the user, raise an error.
+- If `rng` is passed by keyword, validate it with `np.random.default_rng()` and use it to generate random numbers as needed.
+- If `random_state`/`seed`/`...` is specified (by keyword or position, if allowed), preserve existing behavior.
+
+**After `rng` becomes available** in all releases within the support window suggested by SPEC 0, emit warnings as follows:
+
+- If neither `rng` nor `random_state`/`seed`/`...` is specified and `np.random.seed` has been used to set the seed, emit a `FutureWarning` about the upcoming change in behavior.
+- If `random_state`/`seed`/`...` is passed by keyword or by position, treat it as before, but:
+
+  - Emit a `DeprecationWarning` if passed by keyword, warning about the deprecation of keyword `random_state` in favor of `rng`.
+  - Emit a `FutureWarning` if passed by position, warning about the change in behavior of the positional argument.
+
+**After the deprecation period**, accept only `rng`, raising an error if `random_state`/`seed`/`...` is provided.
+
+By now, the function signature, with type annotations, could look like this:
+
+```python
+from collections.abc import Sequence
+import numpy as np
+
+
+SeedLike = int | np.integer | Sequence[int] | np.random.SeedSequence
+RNGLike = np.random.Generator | np.random.BitGenerator
+
+
+def my_func(rng: RNGLike | SeedLike | None = None):
```

Comment:
BTW, is there a good reference for the name "Hinsen principle"? The idea is great, but this is the Google search result for the "famous" Hinsen principle:

![image](https://github.com/user-attachments/assets/bf24f5f0-5cac-42a3-9bdc-756e8f6ed01c)

Reply:
I also had to google for it, but then found a couple of interesting reads that now populates a few browser tabs set aside to be read later :)

Reply:
I suspect it comes from conversations with @khinsen, or perhaps his blog post https://blog.khinsen.net/posts/2017/11/16/a-plea-for-stability-in-the-scipy-ecosystem.html

Konrad complains about the scientific Python ecosystem a fair bit, especially about how we don't pay attention to backward compatibility. So, I suppose it is ironic that we spend so much time on this problem, and only fitting that we name one of our biggest headaches after him ;) 

[Konrad doesn't seem like someone who'd take offense at my humor, but I'll make clear that that's what it is either way.]

Reply:
> Unless switching the behavior in-place is actually intended like the decorator implies

The decorator includes as a comment:
```
                # Note that this decorator does not deprecate positional use of the
                # argument; it only warns that the behavior will change in the future.
                # Simultaneously transitioning to keyword-only use is another option.
```
and the SPEC notes:

> Our recommendation here is a deprecation strategy which does not in _all_ cases adhere to the Hinsen principle[^hinsen], although it could very nearly do so by enforcing the use of `rng` as a keyword argument.

So the decorator does not mean to imply that this is the way to go.

When I first started generalizing the decorator, which worked only if `random_state` (or other old name) was already keyword-only, I had in mind that it would make `rng` keyword-only regardless of whether `random_state` was keyword-only or not. Some considerations that changed the plan to continue to allow positional use:

- Even though I think both of them support making arguments keyword-only, Stefan and Pampile mentioned that the SPEC should not be overly-prescriptive about that. I agree that I wouldn't want this to be held up by a debate about the merits of transitioning to keyword-only arguments; it's complicated enough as it is!
- If `random_state` (or other old name) currently allows positional use, deprecating positional use of `random_state` must be accompanied by deprecating positional use of all following positional arguments. That is a bit more of a disruption.
- There are some other complications depending on what we want the documented signature to look like:
  - If we also want the documented signature to change immediately (e.g. so auto-completion suggests the new signature) as we do now, the decorator has to pass arguments to the underlying function by keyword, which means that the decorator has to know their keyword names, requiring either more complex decorator configuration or use of introspection. 
  - If we _don't_ change the documented signature immediately, we need some sort of admonitions in the documentation about the deprecation of the keywords, which is also going to be more complex.

In short, I took what appeared to be the safer route, but projects are welcome to make it keyword-only at the same time if they wish.

Reply:
> Konrad complains about the scientific Python ecosystem a fair bit, especially about how we don't pay attention to backward compatibility. So, I suppose it is ironic that we spend so much time on this problem, and only fitting that we name one of our biggest headaches after him ;)

Well, I am an academic, any citation is good for me ;-)

It's just a bit sad that Google considers me a typo for a brand of TV sets. But those are the laws of the market...

Reply:
> Well, I am an academic, any citation is good for me ;-)

:joy: 

Since we have you here, Konrad, would you say the above is the best citation of yours to the principle? I can add the URL to the SPEC and to the skimage discussion, so people will hopefully find it more easily in the future.

Alternatively, we can do something a tiny bit more ambitious and write a blog post at https://blog.scientific-python.org to explain the problem.

I don't know if anyone has been following, but scikit-image has been in a major holding pattern over this for years now. We're trying terribly hard to get a new version of the package out, with API designs we consider superior, but in many cases run into silent backward incompatibility. Our only viable solution so far is to release into an entirely new namespace, `skimage2`, which in itself is a headache. We haven't proceeded with that yet, because we are somewhat terrified of estranging existing users, but at the same time know we have to move forward.

Reply:
I should also probably apologize to @khinsen for using the name Scientific Python without his permission (ref https://pypi.org/project/ScientificPython/). An oversight at the time, but hopefully we will receive forgiveness, given that we're trying to fulfill a mission I believe he is strongly in favor of.

Reply:
@mdhaber, I'm not clear on your view on the current change to recommend keyword-only here. Is that okay or would you prefer to revert it before we merge this?

Reply:
I think keyword-only is preferable, too. If the original argument was keyword-only, I think the decorator can keep it keyword-only, so this is not necessarily inconsistent with what the decorator can do. 

I'll think about extending it to make it able to deprecate positional use of an argument if desired. But that doesn't need to hold anything up.

Reply:
> would you say the above is the best citation of yours to the principle?

Yes!

> Our only viable solution so far is to release into an entirely new namespace, skimage2, which in itself is a headache.

I see why it's a headache, but I also believe it's the best solution in the long run. Explicit is better than implicit. New API, new name. And it's [done elsewhere](http://blog.ezyang.com/2016/12/thoughts-about-spec-ulation-rich-hickey/) with good success.

> hopefully we will receive forgiveness, given that we're trying to fulfill a mission I believe he is strongly in favor of.

Indeed I am! The name clash has regularly been a source of confusion, but nothing serious. And by now, my ScientificPython has disappeared from most people's radar, being Python-2-only.